src/book/book.service.ts

```diff
@@ -0,0 +1,144 @@
+import {
+  ConflictException,
+  Injectable,
+  InternalServerErrorException,
+  NotFoundException,
+} from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Author } from 'src/author/author.entity';
+import { Category } from 'src/category/category.entity';
+import { Message } from 'src/error/message-eror';
+import { getConnection } from 'typeorm';
+import { Book } from './book.entity';
+import { BookRepository } from './book.repository';
+import { BookDto } from './dto/book-dto';
+import { authorIsExist, categoryIsExist } from './func/check-exist';
+import { getDateNow } from './func/get-date';
+
+@Injectable()
+export class BookService {
+  constructor(
+    @InjectRepository(BookRepository)
+    private bookRepository: BookRepository,
+  ) {}
+
+  async getBooks(searchBooks: string): Promise<Book[]> {
+    return this.bookRepository.getBook(searchBooks);
+  }
+
+  async getBookById(id: string): Promise<Book> {
+    const found = await this.bookRepository.findOne({
+      where: {
+        id: id,
+        isDelete: false,
+      },
+    });
+    if (!found) {
+      throw new NotFoundException(`${Message.NOT_FOUND.user}=${id}`);
+    }
+    return found;
+  }
+
+  async createBook(newBook: BookDto): Promise<Book> {
+    const {
+      title,
+      authorId,
+      categoryId,
+      publishYear,
+      price,
+      description,
+      cover,
+    } = newBook;
+    try {
+      const categoryFound = await categoryIsExist(categoryId);
+      const authorFound = await authorIsExist(authorId);
+      if (!categoryFound)
+        throw new NotFoundException(Message.NOT_EXIST.cateogory);
```

Comment:
The logic should be like this.
- The body of the request will have categoryName and authorName
- The backend will check if there are category or author exist with that name
- If they are exist, save book with their relations
- If they are not exist create a new one with the name.